pkg/noun/retrieve.c

```diff
@@ -1957,29 +1957,35 @@ u3r_word_buffer(u3_atom* a, c3_w* len_w)
   return pug_u->buf_w;
 }
 
+static inline c3_ys
+_comp_words(c3_w a_w, c3_w b_w)
+{
+  return (c3_ys)(a_w > b_w) - (c3_ys)(a_w < b_w);
+}
+
 /* u3r_comp(): compares two atoms:
 ** returns 1 if a > b, -1 if a < b, 0 if they are equal
 */
 c3_ys
 u3r_comp(u3_atom a, u3_atom b)
 {
-  if (c3y == u3a_is_cat(a) || c3y == u3a_is_cat(b)) {
-    return (c3_ys)(a > b) - (c3_ys)(a < b);
-  }
+  if (c3y == u3a_is_cat(a) || c3y == u3a_is_cat(b)) 
+    return _comp_words(a, b);
+  
   if ( a == b ) return 0;
 
   u3a_atom* a_u = u3a_to_ptr(a);
   u3a_atom* b_u = u3a_to_ptr(b);
 
-  if (a_u->len_w != b_u->len_w) {
-    return (c3_ys)(a_u->len_w > b_u->len_w) - (c3_ys)(a_u->len_w < b_u->len_w);
-  }
+  if (a_u->len_w != b_u->len_w)
+    return _comp_words(a_u->len_w, b_u->len_w);
 
   c3_w* a_w = a_u->buf_w;
   c3_w* b_w = b_u->buf_w;
+
   for (c3_w i_w = a_u->len_w; i_w--;) {
-    if (a_w[i_w] > b_w[i_w]) return 1;
-    if (a_w[i_w] < b_w[i_w]) return -1;
+    if ( a_w[i_w] != b_w[i_w] )
+      return _comp_words(a_w[i_w], b_w[i_w]);
```

Comment:
Style nit: c3 style always includes braces, even if the enclosed statement is only one line.